dali/operators/image/mask/grid_mask.cc

```diff
@@ -0,0 +1,87 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/core/static_switch.h"
+#include "dali/operators/image/mask/grid_mask.h"
+#include "dali/kernels/mask/grid_mask_cpu.h"
+#include "dali/core/format.h"
+
+#define TYPES (uint8_t, int16_t, int32_t, float)
+
+namespace dali {
+
+DALI_SCHEMA(GridMask)
+    .DocStr(R"(Zeroes out pixels of an image in a grid-like fashion. The grid
+consists of squares repeating in x and y directions, with the same spacing in
+both directions. Can be rotated around the origin.)")
+    .NumInput(1)
+    .NumOutput(1)
+    .AddOptionalArg("tile", R"(The length of a single tile, which is equal to
+width of black squares plus the spacing between them.)",
+                    100, true)
+    .AddOptionalArg("ratio",
+                    "The ratio between black square width and tile width.",
+                    0.5f, true)
+    .AddOptionalArg("angle",
+                    "Angle, in radians, by which the grid is rotated.",
+                    0.0f, true);
+
+bool GridMaskCpu::SetupImpl(std::vector<OutputDesc> &output_desc,
+                            const workspace_t<CPUBackend> &ws) {
+  const auto &input = ws.template InputRef<CPUBackend>(0);
+  const auto &output = ws.template OutputRef<CPUBackend>(0);
+  output_desc.resize(1);
+  output_desc[0] = {input.shape(), input.type()};
+  kernel_manager_.Resize(num_threads_, max_batch_size_);
+  TYPE_SWITCH(input.type().id(), type2id, Type, TYPES, (
+      {
+          using Kernel = kernels::GridMaskCpu<Type>;
+          kernel_manager_.Initialize<Kernel>();
+      }
+  ), DALI_FAIL(make_string("Unsupported input type: ", input.type().id()))) // NOLINT
+  return true;
+}
+
+
+void GridMaskCpu::RunImpl(workspace_t<CPUBackend> &ws) {
+  const auto &input = ws.template InputRef<CPUBackend>(0);
+  auto &output = ws.template OutputRef<CPUBackend>(0);
+  output.SetLayout(input.GetLayout());
+  auto out_shape = output.shape();
+  auto& tp = ws.GetThreadPool();
+  TYPE_SWITCH(input.type().id(), type2id, Type, TYPES, (
+      {
+          using Kernel = kernels::GridMaskCpu<Type>;
+          auto in_view = view<const Type>(input);
+          auto out_view = view<Type>(output);
+          for (int sid = 0; sid < input.shape().num_samples(); sid++) {
+            int tile = spec_.GetArgument<int>("tile", &ws, sid);
+            float ratio = spec_.GetArgument<float>("ratio", &ws, sid);
+            float angle = spec_.GetArgument<float>("angle", &ws, sid);
+            auto in_sample = in_view[sid];
+            auto out_sample = out_view[sid];
+            tp.AddWork([&, sid](int tid) {
```

Comment:
You should be careful what you capture by the reference as many variables have very limited scope here - like `tile`, so in the next loop iteration it won't be there, not to mention `tp.RunAll();` line.

Reply:
I think I fixed it, not entirely sure though. Could you check again please?

Reply:
I think it looks good now.